pkg/trace/agent/agent.go

```diff
@@ -110,6 +111,9 @@ func (a *Agent) Run() {
 	info.UpdatePreSampler(*a.Receiver.PreSampler.Stats()) // avoid exposing 0
 
 	a.start()
+	for i := 0; i < runtime.NumCPU()*2; i++ {
```

Comment:
Is `a.work()` a CPU-only operation or does it also wait on some IO?
If that's mostly CPU, `runtime.NumCPU()` should be enough, no reason to have more goroutines (and why would it be 2x? why not 5x? 10x?), except to deal with heterogeneous input (traces taking much longer to process than others).

Reply:
Yep, I'll get rid of the `2`. It was a misunderstanding on my behalf, related to number of physical vs logical cores.

Reply:
My only suggestion here (as per other thread) is to keep this concurrency level configurable (or under feature flag). The rationale is we can ship this functionality without changing the current behavior, giving us way more time to continue deep testing while 6.12 is out.

Then, in 6.13 we can make it on by default using the max number of CPUs.

Reply:
Fair to me, we could introduce a boolean option that make it use `NumCPU` instead of `1`.
It should not be documented and it will be entirely dropped in the release after.

That's just to give us a release to run with parallelism in various environments, catching up potential regressions or race conditions (even if I'm confident, that's the right way to do it).

Reply:
@djova had a 3rd suggestion where we completely remove the `Out` channel and just do the processing in the handler directly. This means that each handler (and in turn, request) gets its own processing goroutine. Even though this sounds crazy, with the coming presampler work that will keep memory and CPU in check (follow up to #3239), this might just be a good approach to try.

And if we are to go super safe behind a feature flag anyway, I'd say the above is worthwhile considering.

Reply:
@LotharSee:

> Fair to me, we could introduce a boolean option that make it use NumCPU instead of 1.
It should not be documented and it will be entirely dropped in the release after.

I'll suggest to add a new yaml option `feature_flags` which guarantees no backward compatibility in terms of values and also has a corresponding env. flag. Might be good moving forward for future moves like this too.

Reply:
To me it sounds good. What is important is that we remove the feature flag on 6.13. I will be way more confident if we proceed with this approach. Thanks!

Reply:
- For the processing in the handler, that's a much bigger design change, happy to discuss it on a dedicated thread?
- I also expect that flag to disappear completely in 2 releases so no need to be BC compatible (the temporary option would then just be ignored in favor of the new global default). So no need to introduce a generic feature flags system (up to you!).

Reply:
Ok, let's keep it simple for now and go forward with this change. If we need to think about alternatives we can do so once we need further performance improvements. Start simple.

Reply:
Yes I would say to keep it simple without increasing the surface of our problems (so the handler). We can discuss it further if we prove this change is not good enough.

Reply:
> @djova had a 3rd suggestion where we completely remove the `Out` channel and just do the processing in the handler directly.

On second thought I think your approach of one worker goroutine per CPU is nicer and easier to reason about. Since it's CPU-bound work it makes sense to limit it to a set number of workers.